### PR TITLE
copy prometheus receiver to generate present metrics

### PR DESCRIPTION
**Pull request by paivagustavo**
This PR adds a Prometheus Discovery Receiver which is responsible to discovery services and their metadata, generating a `present` metric that could be used to:

1. evaluate an `up` metric in conjunction of another possible metric `alive`;
2. enhance metrics and spans with metadata of a service discovered via the Prometheus SD.

The receiver works with the same configuration as the prometheus scrape config, starts the service discovery to get a list of group targets. A group target represents multiples instances of a service, that is expected to be up. With this we generate a `ResourceMetric` for each target. The metadata is added as labels to the `Resource` and inclunding a `job`and `instance` label to uniquely identify a service.

### CODE REVIEW

File: receiver/prometheusdiscoveryreceiver/README.md [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+
+```yaml
+receivers:
+  prometheus:
+    config:
+```
+
+For example:
+
+```yaml
```
Comment (carlosalberto) [inline receiver/prometheusdiscoveryreceiver/README.md:46]: Update this to reflect this receiver configuration? Not a biggie but handy ;)

File: examples/local/otel-config.yaml [hunk 2]
```diff
@@ -42,7 +51,7 @@ service:
       processors: [batch]
       exporters: [logging]
     metrics:
-      receivers: [otlp, opencensus, prometheus]
+      receivers: [otlp, opencensus, prometheus, prometheus_discovery]
```
Comment (carlosalberto) [inline examples/local/otel-config.yaml:54]: Do we need to keep on running `prometheus`? I'd rather leave it out (unless I'm **very** mistaken and we still need it running on the side)

File: receiver/prometheusdiscoveryreceiver/discovery_receiver.go [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+						// The target address is used to identify an instance.
+						tAttrs.InsertString("instance", string(value))
+					} else {
+						tAttrs.InsertString(cleanLabelName(string(name)), string(value))
+					}
+				}
+
+				tAttrs.InsertString("job", job)
+				tAttrs.InsertString("source", "prometheus_discovery")
+				tAttrs.InsertString("instance", string(target["__address__"]))
```
Comment (carlosalberto) [inline receiver/prometheusdiscoveryreceiver/discovery_receiver.go:124]: Is this expected? aren't you trying to set `instance` twice, first in the loop above, and now here again?

File: examples/local/otel-config.yaml (hunk 2 above)
Comment (paivagustavo) [reply examples/local/otel-config.yaml:54]: The `prometheus` receiver scrapes metrics and the `prometheus_discovery` find targets and generate `present` metrics, so we need to keep them both.

File: receiver/prometheusdiscoveryreceiver/discovery_receiver.go (hunk 3 above)
Comment (paivagustavo) [reply receiver/prometheusdiscoveryreceiver/discovery_receiver.go:124]: Nope, forgot to remove it :)

File: receiver/prometheusdiscoveryreceiver/README.md (hunk 1 above)
Comment (paivagustavo) [reply receiver/prometheusdiscoveryreceiver/README.md:46]: updated! :)

File: examples/local/otel-config.yaml (hunk 2 above)
Comment (jmacd) [reply examples/local/otel-config.yaml:54]: Yeah, I see these as independent options. There's probably not much duplication of code, at this point.

File: receiver/prometheusdiscoveryreceiver/internal/logger.go [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+	"go.uber.org/zap"
+)
+
+const (
+	levelKey = "level"
+	msgKey   = "msg"
+)
+
+// NewZapToGokitLogAdapter create an adapter for zap.Logger to gokitLog.Logger
+func NewZapToGokitLogAdapter(logger *zap.Logger) gokitLog.Logger {
```
Comment (jmacd) [inline receiver/prometheusdiscoveryreceiver/internal/logger.go:29]: ❤️ 

I take it other components might benefit from this! (Sidecar also wrangles go-kit logs a bunch.)

**Review by jmacd — APPROVED** [review]
💯  🥇

**Comment by jmacd** [comment]
Let's merge!
🚢

**Review by carlosalberto — APPROVED** [review]

File: receiver/prometheusdiscoveryreceiver/internal/logger.go (hunk 4 above)
Comment (paivagustavo) [reply receiver/prometheusdiscoveryreceiver/internal/logger.go:29]: I'll not take credits for this, it was copy-pasted from the prometheus receiver. As a side note we should at least make this shared between both receivers.

**Pull request merged**